include/osmium/io/metadata_options.hpp

```diff
@@ -0,0 +1,239 @@
+
+*/
+
+#include <osmium/util/string.hpp>
+
+#include <stdexcept>
+#include <string>
+#include <osmium/osm/object.hpp>
+
+namespace osmium {
+
+    namespace io {
+
+        /**
+         * Describes which metadata fields are available in an OSMObject. This is most often used to
+         * describe which metadata fields are available in the objects of an OSM file or which
+         * metadata information should be written to an OSM file.
+         */
+        class metadata_options {
+
+            enum options : char {
+                md_none      = 0x00,
+                md_version   = 0x01,
+                md_timestamp = 0x02,
+                md_changeset = 0x04,
+                md_uid       = 0x08,
+                md_user      = 0x10,
+                md_all       = 0x1f
+            } m_options = md_all;
+
+        public:
+
+            metadata_options() noexcept = default;
+
+            explicit metadata_options(const std::string& attributes) {
+                if (attributes.empty() || attributes == "all" || attributes == "true" || attributes == "yes") {
+                    return;
+                }
+                if (attributes == "none" || attributes == "false" || attributes == "no") {
+                    m_options = options::md_none;
+                    return;
+                }
+
+                const auto attrs = osmium::split_string(attributes, '+', true);
+                int opts = 0;
+                for (const auto& attr : attrs) {
+                    if (attr == "version") {
+                        opts |= options::md_version;
+                    } else if (attr == "timestamp") {
+                        opts |= options::md_timestamp;
+                    } else if (attr == "changeset") {
+                        opts |= options::md_changeset;
+                    } else if (attr == "uid") {
+                        opts |= options::md_uid;
+                    } else if (attr == "user") {
+                        opts |= options::md_user;
+                    } else {
+                        throw std::invalid_argument{std::string{"Unknown OSM object metadata attribute: '"} + attr + "'"};
+                    }
+                }
+                m_options = static_cast<options>(opts);
+            }
+
+            /// At least one metadata attribute should be stored.
+            bool any() const noexcept {
+                return m_options != 0;
+            }
+
+            /// All metadata attributes should be stored.
+            bool all() const noexcept {
+                return m_options == options::md_all;
+            }
+
+            /// No metadata attributes should be stored.
+            bool none() const noexcept {
+                return m_options == 0;
+            }
+
+            bool version() const noexcept {
+                return (m_options & options::md_version) != 0;
+            }
+
+            void set_version(bool flag) noexcept {
+                if (flag) {
```

Comment:
Can we do these without the conditional? I just checked this on https://godbolt.org/ and only the newest GCCs can optimize away the conditional.